src/v2/controllers/vdeployment_controller/exec/reconciler.rs

```diff
@@ -0,0 +1,663 @@
+//    We may not need to support this if we can prove this doesn't happen for our controller.
+pub fn reconcile_core(vd: &VDeployment, resp_o: Option<Response<VoidEResp>>, state: VDeploymentReconcileState) -> (res: (VDeploymentReconcileState, Option<Request<VoidEReq>>))
+    requires vd@.well_formed(),
+    ensures (res.0@, res.1.deep_view()) == model_reconciler::reconcile_core(vd@, resp_o.deep_view(), state@),
+{
+    let namespace = vd.metadata().namespace().unwrap();
+    match state.reconcile_step {
+        VDeploymentReconcileStep::Init => {
+            let req = KubeAPIRequest::ListRequest(KubeListRequest {
+                api_resource: VReplicaSet::api_resource(),
+                namespace: namespace,
+            });
+            let old_vrs_list = Vec::<VReplicaSet>::new();
+            // why this assertion is required
+            assert(old_vrs_list@.map_values(|vrs: VReplicaSet| vrs@) == Seq::<VReplicaSetView>::empty());
+            let state_prime = VDeploymentReconcileState {
+                reconcile_step: VDeploymentReconcileStep::AfterListVRS,
+                new_vrs: None,
+                old_vrs_list: old_vrs_list,
+            };
+            return (state_prime, Some(Request::KRequest(req)))
+        },
+        VDeploymentReconcileStep::AfterListVRS => {
+            if !(is_some_k_list_resp!(resp_o) && extract_some_k_list_resp_as_ref!(resp_o).is_ok()) {
+                return (error_state(state), None);
+            }
+            let objs = extract_some_k_list_resp!(resp_o).unwrap();
+            let vrs_list_or_none = objects_to_vrs_list(objs);
+            if vrs_list_or_none.is_none() {
+                return (error_state(state), None);
+            }
+            let (new_vrs_list, old_vrs_list) = filter_old_and_new_vrs(filter_vrs_list(vrs_list_or_none.clone().unwrap(), vd), vd);
+            // no .last().cloned() in verus because "The verifier does not yet support the following Rust feature: overloaded deref"
+            let state = VDeploymentReconcileState {
+                reconcile_step: VDeploymentReconcileStep::Error,
+                new_vrs: None,
+                old_vrs_list: old_vrs_list.clone(),
+            };
+            assert(state@ == model_reconciler::VDeploymentReconcileState {
+                reconcile_step: VDeploymentReconcileStepView::Error,
+                new_vrs: None,
+                old_vrs_list: state.old_vrs_list@.map_values(|vrs: VReplicaSet| vrs@),
+            });
+            if new_vrs_list.len() == 0 {
+                // no new vrs, create one
+                let (state_prime, req) = create_new_vrs(state, vd);
+                assert(state_prime@ == model_reconciler::create_new_vrs(state@, vd@).0);
+                assert(req.deep_view() == model_reconciler::create_new_vrs(state@, vd@).1);
+                assert(state_prime@ == model_reconciler::reconcile_core(vd@, resp_o.deep_view(), state@).0) by {
```

Comment:
...and here you state that state_prime@ is the same as feeding the state@ to reconcile_core at model level. I don't think this assertion makes sense.